scripts/subspace_learning_lib.py

```diff
@@ -0,0 +1,243 @@
+compute the loss. The steps can be summarized as follows
+        w(theta) = w_init + A * theta
+1. Project theta_subspace ∈ R^d => theta ∈ R^D
+2. Reconstruct the pytree of the reconstructed weights
+3. Compute loss of the model w.r.t. theta_subspace
+
+Author: Gerardo Durán-Martín (@gerdm), Aleyna Kara(@karalleyna)
+'''
+
+import superimport
+
+import numpy as np
+from functools import partial
+
+import jax
+import jax.numpy as jnp
+from jax.random import split, normal
+from jax import vmap, jit, value_and_grad, tree_map
+from jax.lax import scan
+
+from blackjax import nuts, stan_warmup
+from sgmcmcjax.util import progress_bar_scan
+
+
+
+@jit
+def convert_params_from_subspace_to_full(params_subspace, projection_matrix, params_full_init):
+    """
+    Project the subspace model weights onto the full model weights
+    Parameters
+    ----------
+    params_subspace: jnp.ndarray(d)
+        The subspace model weights
+    projection_matrix: jnp.ndarray(D,d)
+        The projection matrix
+    params_full_init: jnp.ndarray(D)
+        The initial full model weights
+    Returns
+    -------
+    params_full: jnp.ndarray(D)
+    """
+    params_full = jnp.matmul(params_subspace, projection_matrix)[0] + params_full_init
+    return params_full
+
+
+def generate_random_basis(key, d, D):
+    projection_matrix = normal(key, shape=(d, D))
+    projection_matrix = projection_matrix / jnp.linalg.norm(projection_matrix, axis=-1, keepdims=True)
+    return projection_matrix
+
+
+@jit
+def adam_update(grads, params, mass, velocity, hyperparams):
+    mass = hyperparams["beta_1"] * mass + (1.0 - hyperparams["beta_1"]) * grads
+    velocity = hyperparams["beta_2"] * velocity + (1.0 - hyperparams["beta_2"]) * (grads ** 2.0)
+    # Bias correction
+    hat_mass = mass / (1 - hyperparams["beta_1"])
+    hat_velocity = velocity / (1 - hyperparams["beta_2"])
+    # Update
+    params = params - hyperparams["lr"] / (jnp.sqrt(hat_velocity) + hyperparams["epsilon"]) * hat_mass
+    return params, mass, velocity
+
+
+def loglikelihood(params, x, y, predict_fn):
+    """Computes the log-likelihood."""
+    logits = predict_fn(params, x)
+    ll = jnp.sum(logits[y])
+    return ll
+
+
+@jit
+def logprior(params_full, prior_variance=1.):
+    """Computes the Gaussian prior log-density."""
+    exp_term = -params_full ** 2 / (2 * prior_variance)
+    norm_constant = 0.5 * jnp.log((2 * jnp.pi * prior_variance))
+    return (exp_term - norm_constant).mean()
+
+
+def accuracy(params, batch, predict_fn):
+    inputs, targets = batch
+    logits = predict_fn(params, inputs)
+    predicted_class = jnp.argmax(logits, axis=-1)
+    return jnp.mean(predicted_class == targets), jnp.exp(logits)
+
+
+def data_stream(dataset, batch_size):
+    X, y = dataset
+    num_train = len(X)
+    num_batches, rem = divmod(num_train, batch_size)
+    num_batches += 1 if rem else 0
+    rng = np.random.RandomState(0)
```

Comment:
I believe it's considered bad practice to mix np.rng and jax.prng. I suggest passing in a jax.prng key.